tests/aws/services/cloudwatch/test_cloudwatch.py

```diff
@@ -2211,6 +2227,51 @@ def log_group_exists():
         )
         snapshot.match("lambda-alarm-invocations", invocation_res)
 
+    @markers.aws.validated
+    def test_get_metric_with_no_results(self, snapshot, aws_client):
+        utc_now = datetime.now(tz=timezone.utc)
+        namespace = f"n-{short_uid()}"
+        metric = f"m-{short_uid()}"
+
+        aws_client.cloudwatch.put_metric_data(
+            Namespace=namespace,
+            MetricData=[
+                {
+                    "MetricName": metric,
+                    "Value": 1,
+                }
+            ],
+        )
+
+        def validate():
+            data = aws_client.cloudwatch.get_metric_data(
+                MetricDataQueries=[
+                    {
+                        "Id": "result",
+                        "MetricStat": {
+                            "Metric": {
+                                "Namespace": namespace,
+                                "MetricName": metric,
+                                "Dimensions": [
+                                    {
+                                        "Name": "foo",
+                                        "Value": "bar",
+                                    }
+                                ],
+                            },
+                            "Period": 60,
+                            "Stat": "Sum",
+                        },
+                    }
+                ],
+                StartTime=utc_now - timedelta(seconds=60),
+                EndTime=utc_now + timedelta(seconds=60),
+            )
+            snapshot.add_transformer(snapshot.transform.key_value("Label"))
+            snapshot.match("result", data)
+
+        retry(validate, sleep_before=(2 if is_aws_cloud() else 0))
```

Comment:
This test is a bit tricky, because we try to verify the negative test case, where no data is expected to be returned. For this kind of tests it's always hard to know how long we should wait (as we expect the data to be unchanged, e.g. always return empty).

With the current implementation, the `retry` doesn't do anything because we don't have a validation in there. 

I would suggest improving this test by also adding a `get-metric-data` query where the metric would be returned (verify this first), and then we can check that the other `get-metric-data` call with the dimensions does not return the result.

Reply:
Thanks for the comment. I went with `list-metrics` op to check when the test metrics appear and then do the snapshot.